File: 백준/Gold/1068. 트리/트리.java
import java.io.*;
import java.util.*;

public class Main {
	static class Node {
		int id;
		Node parent;
		List<Node> children;
		
		Node(int id) {
			this.id = id;
			this.children = new ArrayList<>();
		}
	}
	static int N;
	static int[] parent;
	static Node root;
	static Node[] node;
	static int dfs(Node cur) {
		if(cur.children.isEmpty()) return 1;
		int result = 0;
		for(Node nxt : cur.children) {
			result += dfs(nxt);
		}
		return result;
	}
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = null;
		N = Integer.parseInt(br.readLine());
		parent = new int[N];
		node = new Node[N];
		for(int i = 0; i < N; i++) {
			node[i] = new Node(i);
		}
		st = new StringTokenizer(br.readLine(), " ");
		for(int i = 0; i < N; i++) {
			parent[i] = Integer.parseInt(st.nextToken());
			if(parent[i] == -1) {
				root = node[i];
			} else {
				node[i].parent = node[parent[i]];
				node[parent[i]].children.add(node[i]);
			}
		}
		int num = Integer.parseInt(br.readLine());
		if(num == root.id) {
			System.out.println(0);
		} else {			
			node[num].parent = null;
			node[parent[num]].children.remove(node[num]);
			parent[num] = -1;
			System.out.println(dfs(root));
		}
		br.close();
	}
}

File: 백준/Gold/1068. 트리/트리.cc
#include <iostream>
#include <vector>
#include <algorithm>
#define MAX 51

using namespace std;

struct Node {
    int id;
    vector<int> children;
};

int N, R, D;
Node node[MAX];
int parent[MAX];

void init() {
    cin >> N;
    for (int i = 0; i < N; i++) {
        cin >> parent[i];
        node[i].id = i;
        if (parent[i] == -1) {
            R = i;
        } else {
            node[parent[i]].children.push_back(i);
        }
    }
}

int dfs(int cur) {
    if (cur == D) return 0;
    if (node[cur].children.empty()) return 1;
    int result = 0;
    for (int child : node[cur].children) {
        result += dfs(child);
    }
    return result;
}

void solution() {
    cin >> D;
    if (D == R) {
        cout << 0 << '\n';
        return;
    }
    for (int i = 0; i < N; i++) {
        auto& children = node[i].children;
        children.erase(remove(children.begin(), children.end(), D), children.end());
    }
    cout << dfs(R) << '\n';
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    init();
    solution();
    return 0;
}
